### ISSUE TITLE
Simplify producing trivial branches

### ISSUE THREAD

**Issue by cscherrer** [hakaru-dev/hakaru#81]
Using the code below as input, the code produced by `simplify` includes trivial branches such as
```
match not((kkH == z[n両])): 
  true: 0
  false: 0
```

Here's the input code:
```
fn topic_prior array(prob): 
 fn word_prior array(prob): 
  fn numDocs nat: 
   fn w array(nat): 
    fn doc array(nat): 
     fn z array(nat): 
      fn wordUpdate nat: 
        categorical(array zNew丑 of size(topic_prior): 
          (nat2prob((summate n両 from 0 to size(w): 
            (match (n両 == wordUpdate): 
              true: 0
              false: 
                (match ((zNew丑 == z[n両]) && (w[wordUpdate] == w[n両])): 
                  true: 1
                  false: 0)))
            + 
            1
          ))
          
          
          / 
          (nat2prob((summate n両 from 0 to size(w): 
            (match (n両 == wordUpdate): 
              true: 0
              false: 
                (match (zNew丑 == z[n両]): 
                  true: 1
                  false: 0)))
            + 
            size(word_prior)) 
          )
          
          
          * 
          (nat2prob((summate n両 from 0 to size(w): 
            (match (n両 == wordUpdate): 
              true: 0
              false: 
                (match ((doc[wordUpdate] == doc[n両]) && (zNew丑 == z[n両])): 
                  true: 1
                  false: 0)))
            + 
            1
          ))
        )
```

**Comment by yuriy0** [comment]
This one is a bit tricky. The offending expression never occurs inside of Maple. There is a simplification which does *exactly* this in Maple - but it never gets to see the expression.

What `simplify` returns is:

```
piecewise(n1 = wordUpdate
         ,0
         ,piecewise(kkz2 = idx(z, n1), 1, 0))
```
This gets translated to a "good" series of `match`es (inside the divisor of the first factor).

```
piecewise(n1 = wordUpdate
         ,0
         ,piecewise(And(kkz2 = idx(z, n1),idx(doc, wordUpdate) = idx(doc, n1))
                   ,1
                   ,kkz2 <> idx(z,n1)
                   ,0
                   ,idx(doc, wordUpdate) <> idx(doc, n1)
                   ,0))
```
This translates into a bad series of `match`es. In part, because it is the 'piecewise' representation of a Partition (in which every branch is necessarily disjoint). In making a piecewise from a Partition, we typically pick an arbitrary ordering (any one is fine) - by arbitrary, I mean precisely the linear order which the pieces happen to have. This cuts down on 'mangling' inside of Maple itself.

Note that this case is particularly special, as the result is only 'trivial' because the default branch of a piecewise is given the value [0](https://github.com/hakaru-dev/hakaru/blob/a7096a4e9786b641edfa921457d2143440056b39/haskell/Language/Hakaru/Parser/Maple.hs#L363).

I see at least a few possible solutions to this:

1. A new form of `If` in the Haskell term language which is more amenable to being translated to and from Partition. (Basically, Partition on the Haskell side.)
2. A simplification done in Haskell which normalizes `If` statements to some particular form, up to a certain set of simplifications deemed trivial. 
3. Maple itself tries to put Partitions and piecewise into the form that 'looks' nicest immediately before sending back to Haskell. 
4. Hardcoding this particular case (where the last piece of a piecewise, or any piece of a Partition, is identically `0`). This is essentially a "light" version of 2.

**Comment by JacquesCarette** [comment]
I see no way of quantifying option 3.  It seems quite fragile.

Somehow, it seems that Haskell-side normalization is also going to be fragile, as it really can't do much reasoning.

I see solution 1. as the most viable.  Concurrently, it seems to me that we encounter a huge number of summate-for-counting in our models.  I am increasingly tempted to add a 'counting' primitive to Hakaru to express this pattern.

    counting(n, low, high, predicate)

would count the number of times that `predicate` is true when `n` ranges from `low` to `high-1`.

[Note that Chad's 
![image](https://user-images.githubusercontent.com/1855141/27051174-2909637c-4f83-11e7-8000-43ff868d207b.png)
is incredibly compact because it uses an even shorter notation for `counting`]

**Comment by ccshan** [comment]
@yuriy0, I can't quite understand your explanation. First, you write 'This gets translated to a "good" series' between two code displays, so does 'This' refer to the code display above (what you say '`simplify` returns') or something else, and does 'a "good" series' refer to the code display below or some Hakaru code not shown? Second, which processing step(s) do "gets translated to" and "translates into" refer to?

@JacquesCarette, I'm not sure how adding a `counting` primitive would help, because I don't know how to change our simplification code to produce it. Note also that `Bucket` generalizes `counting`.

**Comment by yuriy0** [comment]
Both "this" refer to the program immediately above. To be abundantly (hopefully!) clear:

The following Maple program fragment
```
piecewise(n1 = wordUpdate
         ,0
         ,piecewise(kkz2 = idx(z, n1), 1, 0))
```
translates to the Hakaru program
```
match (n1 == wordUpdate): 
   true: 0
   false: 
      (match (kkh == z[n1]): 
          true: 1
          false: 0)
```

Whereas the following Maple program fragment
```
piecewise(n1 = wordUpdate
         ,0
         ,piecewise(And(kkz2 = idx(z, n1),idx(doc, wordUpdate) = idx(doc, n1))
                   ,1
                   ,kkz2 <> idx(z,n1)
                   ,0
                   ,idx(doc, wordUpdate) <> idx(doc, n1)
                   ,0))
```
translates to the Hakaru program fragment
```
 match (n1 == wordUpdate): 
    true: 0
    false: 
      (match ((kkh == z[n1]) && (doc[wordUpdate] == doc[n1])): 
         true: 1
         false: 
             (match not((kkh == z[n1])): 
                    true: 0
                    false: 0))
```

Neither Maple program contains `piecewise(.., 0, 0)`, but the latter Hakaru program fragment does.

**Comment by ccshan** [comment]
Thanks @yuriy0, that helps me a lot to understand you! None of the expression you posted contained `Partition`, so I'm not sure how it's involved -- does some Maple expression you posted originate from an expression containing `Partition`? (And if so, which expression containing `Partition`? I need to ask because maybe one `piecewise` originates from one `Partition` or maybe two `piecewise`s together originate from a single `Partition`.)

It seems that @cscherrer's original complaint can be addressed in a not-entirely-fragile way by merging `Partition` branches whose right-hand-sides are equal (or if `Partition` is not actually involved, by merging **adjacent** `piecewise` branches whose right-hand-sides are equal), no?

**Comment by yuriy0** [comment]
> None of the expression you posted contained Partition, so I'm not sure how it's involved

Currently, Partition is converted to piecewise before being sent back to Haskell. 

> does some Maple expression you posted originate from an expression containing Partition?

All of the piecewise in question were at some point in time Partitions. This is probably true of nearly every piecewise encountered (it will be a Partition at some point).

> I need to ask because maybe one piecewise originates from one Partition or maybe two piecewises together originate from a single Partition

One piecewise always originates from one Partition, in the sense that `PartitionToPW`, or any other Parition function, will never split a Partition into multiple piecewise. 

> It seems that @cscherrer's original complaint can be addressed in a not-entirely-fragile way by merging Partition branches whose right-hand-sides are equal

This is done in Partition:-Simpl; and upon another examination of the program, I agree this would fix the problem, and it is likely a better solution than the 4 things I listed before. I'm currently looking into why exactly it doesn't happen here!

**Comment by JacquesCarette** [comment]
@ccshan I think I know how to generate `counting`.  I thought `Bucket` 'generated' an array -- `counting` is a scalar.  I would be quite happy to generate something more general.  `counting` is indeed to restricted, although it sure does seem to come up a lot.

Right now, the communication between Haskell and Maple goes through `piecewise` (and `If`?).  We should definitely try to deprecate that, and communicate more directly through `Partition`.  But that would require adding a `Partition` concept to hakaru; are you ok with that?

**Comment by yuriy0** [comment]
This is seemingly fixed by c35c12f (although it brought to attention some very tangentially related bugs, 1f64a5a89, b5562ff70 and 737f766a3a7 are related). 

The current output is: 
```
(fn topic_prior array(prob):
  (fn word_prior array(prob):
    (fn numDocs nat:
      (fn w array(nat):
        (fn doc array(nat):
          (fn z array(nat):
            (fn wordUpdate nat:
              categorical(array kkh of size(topic_prior):
                           ((nat2prob((summate n1 from 0 to size(w):
                                        (match ((n1 == wordUpdate) ||
                                                not((kkh == z[n1])) ||
                                                not((w[wordUpdate] == w[n1]))):
                                          true: 0
                                          false: 1)))
                              +
                             1)
                             /
                            nat2prob(((summate n1 from 0 to size(w):
                                        (match ((n1 == wordUpdate) || not((kkh == z[n1]))):
                                          true: 0
                                          false: 1))
                                       +
                                      size(word_prior)))
                             *
                            (nat2prob((summate n1 from 0 to size(w):
                                        (match ((n1 == wordUpdate) ||
                                                not((kkh == z[n1])) ||
                                                not((doc[wordUpdate] == doc[n1]))):
                                          true: 0
                                          false: 1)))
                              +
                             1))))))))))
```


The other discussion: 

>  I think I know how to generate counting. I thought Bucket 'generated' an array -- counting is a scalar. I would be quite happy to generate something more general. counting is indeed to restricted, although it sure does seem to come up a lot.

Right now, the communication between Haskell and Maple goes through piecewise (and If?). We should definitely try to deprecate that, and communicate more directly through Partition. But that would require adding a Partition concept to hakaru; are you ok with that?

probably deserve its own issue.
